docs/src/proposals/program-attributes.md

```diff
@@ -0,0 +1,102 @@
+# Program Attributes
+
+## Problem
+
+Instructions are currently defined with an `enum`:
+
+```rust,ignore
+pub enum SystemInstruction {
+    /// Assign account to a program
+    ///
+    /// # Account references
+    ///   0. [WRITE, SIGNER] Assigned account public key
+    Assign {
+        /// Owner program account
+        owner: Pubkey,
+    },
+}
+```
+
+and a constructor:
+
+```rust,ignore
+pub fn assign(pubkey: &Pubkey, owner: &Pubkey) -> Instruction {
+    let account_metas = vec![AccountMeta::new(*pubkey, true)];
+    Instruction::new(
+        system_program::id(),
+        &SystemInstruction::Assign { owner: *owner },
+        account_metas,
+    )
+}
+```
+
+Nearly everything in that constructor is duplicate information, but we
+currently can't generate that constructor from the enum definition because
+the list of account references is in code comments.
+
+## Proposed Solution
+
+Move the data from code comments to attributes, such that the constructors
+can be generated, and include all the documentation from the enum definition.
+
+
+```rust,ignore
+use solana_sdk::{
+    program_macros::{program, accounts, WRITE, SIGNER}
```

Comment:
This is only here to point out why `WRITE` and `SIGNER` were capitalized. They were meant to be imported as constants.  Maybe combine them with a bitwise or instead of the comma-separated list?